azurelinuxagent/common/protocol/wire.py

```diff
@@ -1811,3 +1934,98 @@ def is_on_hold(self):
+            </Plugin>
+            <Plugin name="Microsoft.Flipperdoodle" version="1.0.0">
+                <RuntimeSettings seqNo="0">{"runtimeSettings":[{"handlerSettings":{"protectedSettingsCertThumbprint":"4037FBF5F1F3014F99B5D6C7799E9B20E6871CB3","protectedSettings":"MIICWgYJK","publicSettings":{"foo":"bar"}}}]}</RuntimeSettings>
+            </Plugin>
+        </PluginSettings>
+        <StatusUploadBlob statusBlobType="BlockBlob">https://putblobhere</StatusUploadBlob>
+        </Extensions>
+        """
+        if 'extensionGoalStates' in self.__dict__:
+            import xml.etree.ElementTree as xml
+            root = xml.Element("Extensions")
+            root.set('version', '1.0.0.0')
+            root.set('goalStateIncarnation', str(self.inVMArtifactsProfileBlobSeqNo))
+
+            # Some properties such as state are required and will generate an exception if not there
+            # Others such as autoUpgrade are optional so we'll check first
+            plugins = xml.SubElement(root, 'Plugins')
+            for extensionGoalState in self.extensionGoalStates:
+                plugin = xml.SubElement(plugins, 'Plugin')
+                plugin.set('name', extensionGoalState['name'])
+                plugin.set('version', extensionGoalState['version'])
+                plugin.set('location', extensionGoalState['location'])
+                plugin.set('state', extensionGoalState['state'])
+                plugin.set('autoUpgrade', str(extensionGoalState.get('autoUpgrade')))
+                plugin.set('failoverlocation', str(extensionGoalState.get('failoverlocation')))
+                plugin.set('runAsStartupTask', str(extensionGoalState.get('runAsStartupTask')))
+                plugin.set('isJson', 'true')
+                plugin.set('useExactVersion', str(extensionGoalState.get('useExactVersion')))
+
+            plugin_settings = xml.SubElement(root, 'PluginSettings')
+            for extensionGoalState in self.extensionGoalStates:
+                plugin = xml.SubElement(plugin_settings, 'Plugin')
+                plugin.set('name', extensionGoalState['name'])
+                plugin.set('version', extensionGoalState['version'])
+                runtime_settings = xml.SubElement(plugin, 'RuntimeSettings')
+                runtime_settings.set('seqNo', str(extensionGoalState['settingsSeqNo']))
+
+                # CRP may send multiple settings but this agent only supports one
+                handler_settings = extensionGoalState['settings'][0]
+                json_settings = json.dumps({"runtimeSettings": [{"handlerSettings": handler_settings}]})
+                runtime_settings.text = json_settings
+
+            config_xml = xml.tostring(root).decode()
+            self.save_config(config_xml)
+            extensions_config = ExtensionsConfig(config_xml)
+
+            return extensions_config
+        return None
+
+    def save_config(self, config_xml):
+        # before we write the new file, remove all the old ones
```

Comment:
If you call the archive function here `self.goal_state_flusher.flush(datetime.utcnow())` you dont have to implement a separate file_deletion logic